File: Platforms/Win32/Reign_Video_D3D9_Component/DepthStencilState.cpp
#include "pch.h"
#include "DepthStencilState.h"

namespace Reign_Video_D3D9_Component
{
	#pragma region Constructors
	DepthStencilStateCom::DepthStencilStateCom(VideoCom^ video, DepthStencilStateDescCom^ desc)
	{
		this->video = video;
		this->desc = desc;
	}
	#pragma endregion
	
	#pragma region Methods
	void DepthStencilStateCom::Enable()
	{
		video->device->SetRenderState(D3DRS_ZENABLE, desc->depthReadEnable);
		if (desc->depthReadEnable == D3DZB_TRUE)
		{
			video->device->SetRenderState(D3DRS_ZFUNC, desc->depthFunc);
		}
		video->device->SetRenderState(D3DRS_ZWRITEENABLE, desc->depthWriteEnable);

		video->device->SetRenderState(D3DRS_STENCILENABLE, desc->stencilEnable);
		if (desc->stencilEnable)
		{
			//D3DRS_TWOSIDEDSTENCILMODE
			video->device->SetRenderState(D3DRS_STENCILFUNC, desc->stencilFunc);
			video->device->SetRenderState(D3DRS_STENCILFAIL, desc->stencilFailOp);
			video->device->SetRenderState(D3DRS_STENCILZFAIL, desc->stencilDepthFailOp);
			video->device->SetRenderState(D3DRS_STENCILPASS, desc->stencilPassOp);

			/*video->device->SetRenderState(D3DRS_CCW_STENCILFUNC, desc->stencilFunc);
			video->device->SetRenderState(D3DRS_CCW_STENCILFAIL, desc->stencilFailOp);
			video->device->SetRenderState(D3DRS_CCW_STENCILZFAIL, desc->stencilDepthFailOp);
			video->device->SetRenderState(D3DRS_CCW_STENCILPASS, desc->stencilPassOp);*/
		}
	}
	#pragma endregion
}

File: Platforms/Win32/Reign.Video.D3D9/States/DepthStencilState.cs
using System;
using Reign.Core;
using Reign_Video_D3D9_Component;

namespace Reign.Video.D3D9
{
	public class DepthStencilState : DisposableResource, IDepthStencilState
	{
		#region Properties
		private DepthStencilStateCom com;
		#endregion

		#region Constructors
		public DepthStencilState(IDisposableResource parent, IDepthStencilStateDesc desc)
		: base(parent)
		{
			var video = parent.FindParentOrSelfWithException<Video>();
			com = new DepthStencilStateCom(video.com, ((DepthStencilStateDesc)desc).com);
		}
		#endregion

		#region Methods
		public void Enable()
		{
			com.Enable();
		}
		#endregion
	}
}